File: 1160-letter-tile-possibilities/1160-letter-tile-possibilities.cpp
class Solution {
public:
    void backtrack(string &tiles, vector<bool> &seen, string &curr, unordered_set<string> &st) {
        if (!curr.empty()) st.insert(curr);
        for (int i = 0; i < tiles.size(); i++) {
            if (!seen[i]) {
                seen[i] = true;
                curr.push_back(tiles[i]);
                backtrack(tiles, seen, curr, st);
                curr.pop_back();
                seen[i] = false;
            }
        }
    }

    int numTilePossibilities(string tiles) {
        unordered_set<string> st;
        vector<bool> seen(tiles.size(), false);
        string curr;
        backtrack(tiles, seen, curr, st);
        return st.size();
    }
};

File: 1160-letter-tile-possibilities/1160-letter-tile-possibilities.java
class Solution {
    void func(String tiles,StringBuilder curr,boolean seen[],Set<String>st){
        st.add(curr.toString());
        for(int i=0;i<tiles.length();i++){
            if(!seen[i]){
                seen[i]=true;
                func(tiles,curr.append(tiles.charAt(i)),seen,st);
                curr.deleteCharAt(curr.length()-1);
                seen[i]=false;
            }
        }
    }

    public int numTilePossibilities(String tiles) {
        boolean seen[]=new boolean[tiles.length()];
        Set<String>st=new HashSet<>();
        func(tiles,new StringBuilder(),seen,st);
        return st.size()-1;   
    }
}